### PR TITLE
[Auth] Add user progress tracking - backend

### PR DESCRIPTION
**Pull request by jennifer-lu**
## Notion ticket link

[\[Auth\] Add user progress tracking - backend](https://www.notion.so/uwblueprintexecs/Auth-Add-user-progress-tracking-backend-e3efe0890b0b488d800ecca45298bb04)

## Implementation description
* Added `progress` model (see ticket for details)
* Added progress REST APIs

## Steps to test
1. Open `/pages/index.tsx`
2. Replace the following code (line 195):
```
const Home: React.FC = () => {
    const [showAlert, setShowAlert] = useState(true);
    const [user, setUser] = useState<any | null>(null);

    const variant = useMediaQuery({ query: `(max-width: 925px)` })
        ? "mobile"
        : "desktop";

    useEffect(() => {
        const getCurrentUser = async () => {
            const currentUser = await Auth.currentAuthenticatedUser();
            setUser(currentUser);
        };

        getCurrentUser().catch(console.error);
    }, []);

    return (
        <>
```
With this:
```
const Home: React.FC = () => {
    const [showAlert, setShowAlert] = useState(true);
    const [user, setUser] = useState<any | null>(null);

    const variant = useMediaQuery({ query: `(max-width: 925px)` })
        ? "mobile"
        : "desktop";

    const handleGet = async () => {
        const response = await axios({
            method: "GET",
            url: "/api/progress/get",
            params: {
                username: "testing123",
                module: "621a47377ed5c82f717b2b6d",
            },
        });
        console.log(response);
    };

    const handlePatch = async () => {
        const response = await axios({
            method: "PATCH",
            url: "/api/progress/patch",
            params: {
                username: "testing123",
                module: "621a47377ed5c82f717b2b6d",
            },
            data: {
                username: "testing123",
                module: "621a47377ed5c82f717b2b6d",
                completion: 100,
                input: [],
            },
        });
        console.log(response);
    };

    const handlePost = async () => {
        const response = await axios({
            method: "POST",
            url: "/api/progress/post",
            data: {
                username: "testing123",
                module: "621a47377ed5c82f717b2b6d",
            },
        });
        console.log(response);
    };

    useEffect(() => {
        const getCurrentUser = async () => {
            const currentUser = await Auth.currentAuthenticatedUser();
            setUser(currentUser);
        };

        getCurrentUser().catch(console.error);
    }, []);

    return (
        <>
            <Button onClick={handleGet}>GET</Button>
            <Button onClick={handlePatch}>PATCH</Button>
            <Button onClick={handlePost}>POST</Button>
```
3. Go to a module from the admin dashboard, copy the module id from the url, and replace all instances of "621a47377ed5c82f717b2b6d" in the code above with the id (i.e. replace "621a47377ed5c82f717b2b6d" with a working module id)
4. Save and go to http://localhost:3000/
5. Observe the GET, PATCH, and POST buttons in the top left
6. Open the browser console
7. Click on POST
8. Check that the object logged in the console has method "post" and data containing:
```
completion: 0
createdAt: "2022-03-13T19:18:38.008Z" // time button clicked
module: Array [ "621a47377ed5c82f717b2b6d" ] // your module id
updatedAt: "2022-03-13T19:18:38.008Z" // time button clicked
username: "testing123"
```
9. Click on GET
10. Check that the object logged in the console has method "get" and data containing:
```
completion: 0
createdAt: "2022-03-13T19:18:38.008Z"
module: Array [ "621a47377ed5c82f717b2b6d" ] // your module id
updatedAt: "2022-03-13T19:18:38.008Z"
username: "testing123"
```
11. Click on PATCH
12. Check that the object logged in the console has method "patch" and data containing:
```
completion: 100
createdAt: "2022-03-13T19:18:38.008Z"
module: Array [ "621a47377ed5c82f717b2b6d" ] // your module id
updatedAt: "2022-03-13T19:18:59.048Z" // time button clicked
username: "testing123"
```
13. Click on GET
14. Check that the object logged in the console has method "get" and data containing:
```
completion: 100
createdAt: "2022-03-13T19:18:38.008Z"
module: Array [ "621a47377ed5c82f717b2b6d" ] // your module id
updatedAt: "2022-03-13T19:18:59.048Z"
username: "testing123"
```
15. Feel free to play around with the requests.

## What should reviewers focus on?
* Does it work?
* Does implementation (schema, apis, style, etc.) look ok?

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: database/models/progress.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { model, models, Schema, Types } from "mongoose";
+
+export interface ProgressInterface {
+    _id: Types.ObjectId;
+    username: string;
+    module: string;
+    completion: number;
+    input: Schema.Types.Mixed;
```
Comment (tawnyzhao) [inline database/models/progress.ts:8]: I think the should just be `any` as a type.

Comment (tawnyzhao) [reply database/models/progress.ts:8]: Additionally, it's not `required` in the schema so it should be optional as a type

**Review by tawnyzhao — APPROVED** [review]
Looks good :)

File: pages/api/progress/get.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import { ErrorResponse } from "types/ErrorResponse";
+import { Progress, ProgressInterface } from "../../../database/models/progress";
+import connectDB from "../utils/mongoose";
+
+const get = async (
+    req: NextApiRequest,
+    res: NextApiResponse<ProgressInterface | ErrorResponse>,
+): Promise<void> => {
+    const username = req.query.username as string; // TODO: is this ok
```
Comment (ChinemeremChigbo) [inline pages/api/progress/get.ts:10]: I don't think this is an issue.

**Review by ChinemeremChigbo — APPROVED** [review]
Sweet! Great work I don't think we need to worry about users passing in arrays as query params so the type coercion is fine.

File: database/models/progress.ts (hunk 1 above)
Comment (jennifer-lu) [reply database/models/progress.ts:8]: Using type 'any' doesn't work, so I'm keeping mixed for now (mixed seems to be the standard for mongoose schemas).

**Pull request merged**